cipher/ascon/ascon.go

```diff
@@ -0,0 +1,261 @@
+
+	return c, nil
+}
+
+// NonceSize returns the size of the nonce that must be passed to Seal
+// and Open.
+func (a *Cipher) NonceSize() int { return NonceSize }
+
+// Overhead returns the maximum difference between the lengths of a
+// plaintext and its ciphertext.
+func (a *Cipher) Overhead() int { return TagSize }
+
+// Seal encrypts and authenticates plaintext, authenticates the
+// additional data and appends the result to dst, returning the updated
+// slice. The nonce must be NonceSize() bytes long and unique for all
+// time, for a given key.
+//
+// To reuse plaintext's storage for the encrypted output, use plaintext[:0]
+// as dst. Otherwise, the remaining capacity of dst must not overlap plaintext.
+func (a *Cipher) Seal(dst, nonce, plaintext, additionalData []byte) []byte {
+	if len(nonce) != NonceSize {
+		panic(ErrNonceSize)
+	}
+
+	ptLen := len(plaintext)
+	output := make([]byte, ptLen+TagSize)
+	ciphertext, tag := output[:ptLen], output[ptLen:]
+
+	a.initialize(nonce)
+	a.assocData(additionalData)
+	a.procText(plaintext, ciphertext, true)
+	a.finalize(tag)
+
+	return output
+}
+
+// Open decrypts and authenticates ciphertext, authenticates the
+// additional data and, if successful, appends the resulting plaintext
+// to dst, returning the updated slice. The nonce must be NonceSize()
+// bytes long and both it and the additional data must match the
+// value passed to Seal.
+//
+// To reuse ciphertext's storage for the decrypted output, use ciphertext[:0]
+// as dst. Otherwise, the remaining capacity of dst must not overlap plaintext.
+//
+// Even if the function fails, the contents of dst, up to its capacity,
+// may be overwritten.
+func (a *Cipher) Open(dst, nonce, ciphertext, additionalData []byte) ([]byte, error) {
+	if len(nonce) != NonceSize {
+		panic(ErrNonceSize)
+	}
+	if len(ciphertext) < TagSize {
+		return nil, ErrDecryption
+	}
+
+	ptLen := len(ciphertext) - TagSize
+	plaintext := make([]byte, ptLen)
+	ciphertext, tag0 := ciphertext[:ptLen], ciphertext[ptLen:]
+	tag1 := (&[TagSize]byte{})[:]
+
+	a.initialize(nonce)
+	a.assocData(additionalData)
+	a.procText(ciphertext, plaintext, false)
+	a.finalize(tag1)
+
+	if subtle.ConstantTimeCompare(tag0, tag1) == 0 {
+		return nil, ErrDecryption
+	}
+
+	return plaintext, nil
+}
+
+func (a *Cipher) initialize(nonce []byte) {
+	bcs := blockSize * uint64(a.mode)
+	pB := permB + 2*(uint64(a.mode)-1)
+	a.s[0] = ((KeySize * 8) << 56) | ((bcs * 8) << 48) | (permA << 40) | (pB << 32)
+	a.s[1] = a.key[0]
+	a.s[2] = a.key[1]
+	a.s[3] = binary.BigEndian.Uint64(nonce[0:8])
+	a.s[4] = binary.BigEndian.Uint64(nonce[8:16])
+	a.perm(permA)
+	a.s[3] ^= a.key[0]
+	a.s[4] ^= a.key[1]
+}
+
+func (a *Cipher) assocData(add []byte) {
+	bcs := blockSize * int(a.mode)
+	pB := permB + 2*(int(a.mode)-1)
+	if len(add) > 0 {
+		for ; len(add) >= bcs; add = add[bcs:] {
+			for i := 0; i < bcs; i += 8 {
+				a.s[i/8] ^= binary.BigEndian.Uint64(add[i : i+8])
+			}
+			a.perm(pB)
+		}
+		if len(add) >= 0 {
```

Comment:
This expression is always true, right?

Reply:
unless it's skipped by the if in line 154

Reply:
`len(add) ≥ 0` is always true as a length is always... ≥0!

Reply:
correct, just changed.